materialize-firebolt/transactor.go

```diff
@@ -0,0 +1,287 @@
+}
+
+// firebolt is delta-update only, so loading of data from Firebolt is not necessary
+func (t *transactor) Load(it *pm.LoadIterator, _ <-chan struct{}, _ <-chan struct{}, loaded func(int, json.RawMessage) error) error {
+	panic("delta updates have no load phase")
+}
+
+func (t *transactor) Prepare(_ context.Context, _ pm.TransactionRequest_Prepare) (pf.DriverCheckpoint, error) {
+	checkpoint := FireboltCheckpoint{
+		Queries: *t.queries,
+	}
+	for i, _ := range t.bindings {
+		checkpoint.Files = append(checkpoint.Files, TemporaryFileRecord{
+			Bucket:  t.bucket,
+			Key:     fmt.Sprintf("%s%s.json", t.prefix, randomDocumentKey()),
+			Binding: i,
+		})
+	}
+
+	jsn, err := json.Marshal(checkpoint)
+	if err != nil {
+		return pf.DriverCheckpoint{}, fmt.Errorf("creating checkpoint json: %w", err)
+	}
+	t.checkpoint = checkpoint
+
+	return pf.DriverCheckpoint{DriverCheckpointJson: jsn}, nil
+}
+
+func (t *transactor) projectDocument(spec *pf.MaterializationSpec_Binding, keys tuple.Tuple, values tuple.Tuple) ([]byte, error) {
+	var document = make(map[string]interface{})
+
+	// Add the keys to the document.
+	for i, value := range keys {
+		var propName = spec.FieldSelection.Keys[i]
+		document[propName] = value
+	}
+
+	// Add the non-keys to the document.
+	for i, value := range values {
+		var propName = spec.FieldSelection.Values[i]
+
+		if raw, ok := value.([]byte); ok {
+			document[propName] = json.RawMessage(raw)
+		} else {
+			document[propName] = value
+		}
+	}
+
+	jsonDoc, err := json.Marshal(document)
+	if err != nil {
+		return nil, fmt.Errorf("failed to serialize the addition document: %w", err)
+	}
+
+	return jsonDoc, nil
+}
+
+// write file to S3 and keep a record of them to be copied to the table
+func (t *transactor) Store(it *pm.StoreIterator) error {
+	awsConfig := aws.Config{
+		Credentials: credentials.NewStaticCredentials(t.awsKeyId, t.awsSecretKey, ""),
+		Region:      &t.awsRegion,
+	}
+	sess := session.Must(session.NewSession(&awsConfig))
+	uploader := s3manager.NewUploader(sess)
+
+	uploads := []s3manager.BatchUploadObject{}
+
+	files := make(map[int][]byte)
+
+	for it.Next() {
+		doc, err := t.projectDocument(t.bindings[it.Binding].spec, it.Key, it.Values)
+		if err != nil {
+			return fmt.Errorf("projecting new store json: %w", err)
+		}
+
+		if _, ok := files[it.Binding]; ok {
+			files[it.Binding] = append(files[it.Binding], byte('\n'))
+			files[it.Binding] = append(files[it.Binding], doc...)
+		} else {
+			files[it.Binding] = doc
+		}
+	}
+
+	for bindingIndex, data := range files {
+		cp := t.checkpoint.Files[bindingIndex]
+
+		uploads = append(uploads, s3manager.BatchUploadObject{
+			Object: &s3manager.UploadInput{
+				Bucket: aws.String(cp.Bucket),
+				Key:    aws.String(cp.Key),
+				Body:   bytes.NewReader(data),
+			},
+		})
+	}
+
+	uploadIterator := s3manager.UploadObjectsIterator{
+		Objects: uploads,
+	}
+
+	err := uploader.UploadWithIterator(aws.BackgroundContext(), &uploadIterator)
```

Comment:
One question, though: What happens if Firebolt is asked to load a file that doesn't exist? Is that a problem, or a no-op? In `Prepare`, we don't yet know whether we'll see a given binding in the current transaction and must assume we will and add a file for it to the checkpoint.

Even if firebolt were configurable to not care that a loaded file doesn't exist, we _still_ might want to surface and error on that just for sanity's sake -- to ensure the file wasn't scrubbed accidentally or something.

Thinking out loud, the Flow runtime _does_ know which bindings appear in a transaction, and could send that binding list to Prepare, which could populate checkpoint files appropriately. Then `Store` probably _shouldn't_ use the lazy initialization as I suggest above and instead initialize each stream proactively.

If this is sounding reasonable, perhaps a half-way step is to have Store create files even though they might be empty, and we'll fix it later through a coordinated update with the Flow runtime.

Reply:
Loading a file into main table is done using a query like this:

```
INSERT INTO files SELECT * FROM files_external WHERE source_file_name IN ('file_name.json');
```

If the file doesn't exist, this query is a no-op because the `SELECT ... WHERE` will have no results.

Reply:
I will look into this in a separate PR